### PR TITLE
User config watcher

### PR DESCRIPTION
**Pull request by asymmetric**
This PR adds a watcher for the `user.toml` file.

It's based on @krnowak's still-unmerged work on the file watcher.

### CODE REVIEW

File: components/sup/src/manager/service/mod.rs
Comment (asymmetric) [inline components/sup/src/manager/service/mod.rs]: Wasn't sure how to create a symlink for this. Also, the function is private, so I'm not sure if there's a point.

**Comment by asymmetric** [comment]
**NB**: The config is not currently reloaded, so this doesn't work yet.

File: components/sup/src/manager/mod.rs
Comment (krnowak) [inline components/sup/src/manager/mod.rs]: If you changed the order of members in `Manager` struct, it would be good to reorder the initializations here too.

File: components/sup/src/manager/service/mod.rs
Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: I guess this is a note to yourself, because I don't understand what it is about. Not enough context.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: This comment does not seem to be true - `reset_events_for` sets the boolean to false.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Instead of using `expect`, I'd use `outputln!` macro to report an error and maybe quit the thread. I'm not sure if we should panic in this case.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Why sleeping instead of `wait_for_events`?

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Consider adding this log key to the documentation here - `www/source/docs/reference/log-keys.html.md`.

**Review by krnowak — CHANGES_REQUESTED** [review]
Looks good overall, I had some nits.

File: components/sup/src/manager/service/mod.rs
Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: The `svc_file_path` function could be rendered as a link, since `///` comments are Markdown. I just don't know what the target should be, as it seems to be relative to something.

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Yes, this is a note to myself saying that we could change the implementation. True/false values can be represented by presence/absence in a set.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: One note about this - when you use the `into_path()` function, you also take over the responsibility of removing the created temporary directory. Which you probably don't do, no?

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Returning references makes the trait less flexible, because you need an existing, non-temporary thing in memory to return its address (like a member of a struct that implements this trait). In other words - you can't compute the return value. Not sure if it is really important, but in general I would just return plain values. Of course the downside of returning values is that you need to clone the stuff, but OTOH this won't likely be a bottleneck ever.

**Review by krnowak — CHANGES_REQUESTED** [review]
Another two things I noticed.

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: My rationale here is that the name is always a reference to its containing struct.

Isn't it enough for users of this trait to be able to call `name.clone()` if they want to own the `String`?

Can you show an example of where the current implementation creates issues? As I'm not sure I understand your comment.

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Good point, I've committed a change.

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Good point, fixed.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [reply components/sup/src/manager/user_config_watcher.rs]: There are no problems with current implementation. It was just a general note about returning references in trait functions. So in this case it's likely ok, but in general I would avoid it.

File: components/sup/src/manager/service/mod.rs
Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: Fixed this.

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Tried using a `HashSet` and turns out it's more complicated than expected (if we put a `HashSet` in an `Arc`, that `HashSet` is immutable, unless I use mutexes), so not worth it.

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: OK, let me know if you feel strongly about changing this, otherwise I'm gonna leave it as is.

**Comment by asymmetric** [comment]
@krnowak addressed your comments, but I guess this will just wait until your PR gets merged.

I also rebased on yours.

I will clean up the commit messages next week.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [reply components/sup/src/manager/user_config_watcher.rs]: Yeah, you would need to do something similar to what some code in butterfly does - `Arc<RwLock<HashSet<Whatever>>>`. Anyway, I think that current solution is fine as well.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [reply components/sup/src/manager/user_config_watcher.rs]: Nah, leave it as is, please. If somebody else points out this thing, then you can change it.

File: components/sup/src/manager/service/mod.rs
Comment (krnowak) [inline components/sup/src/manager/service/mod.rs]: Should this `|| self.user_config_updated` be moved one line below, to the `let reconfigure` line?

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: We have an `add` function. Do we need a `remove` function then?

**Review by krnowak — COMMENTED** [review]
I had two questions, otherwise looks good.

**Comment by asymmetric** [comment]
The PR is in a weird state right now. It shows commit that are not in the PR's branch. What happened is that the base branch got merged into master, and after that happened, I rebased my branch on master.

I then changed the PR to use master as the base branch, but now it shows commits here that shouldn't be in this branch (and aren't there according to `git log`).
It seems like this is a problem with GitHub's UI.

Also, please don't be too picky about commit messages (and the commits themselves), as I am still testing this PR out.

File: components/sup/src/manager/service/mod.rs
Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: No, the `needs_reconfiguration` flag is only used to [trigger a `reconfigure` hook](https://github.com/habitat-sh/habitat/blob/c827ae61fa9077974e22fdbf04aecbafefbcef19/components/sup/src/manager/service/mod.rs#L639-L646), which may not be defined. The `reload` flag is actually going to trigger a service reload, which is what we need here.

**Comment by asymmetric** [comment]
@krnowak could you take a look again?

File: components/sup/src/manager/mod.rs
Comment (krnowak) [inline components/sup/src/manager/mod.rs]: To move, not copy. After swapping, the vector under `self.services` lock will be empty.

File: components/sup/src/manager/service/mod.rs
Comment (krnowak) [inline components/sup/src/manager/service/mod.rs]: This became a bit hard to follow. Would pseudo-rust below be clearer? (`cufr` = `cfg_updated_from_rumors`, `sucu` = `self.user_config_updated`, `crc` = `census_ring.changed()`)

```
if sucu {
    if let Err(e) = self.cfg.load_user(&self.pkg) {…}
}
let (cfg_changed, reload, reconfigure) = {
    let ctx = …;
    if cufr || crc {
        (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))
    } else if sucu {
        (true, true, self.compile_configuration(&ctx))
    } else {
        (false, self.needs_reload, self.needs_reconfiguration)
    }
}
self.needs_reload = reload;
self.needs_reconfiguration = reconfigure;
self.user_config_updated = false;
/* return */ cfg_changed
```

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs:64]: That's nice.

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: I'd add "eventually" - "thereby eventually stopping …"

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: thereby is your new favourite word, eh? :)

File: components/sup/src/manager/user_config_watcher.rs
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Drop this empty line?

**Review by krnowak — CHANGES_REQUESTED** [review]
Looks mostly fine, but there was one hairy piece of code I'd like to see if it can be "balder".

File: components/sup/src/manager/service/mod.rs
Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: The `something_to_do` variable would be gone.

Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: I find the 3 value tuple even less clear tbh...

> (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))

what?

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: It's a pretty cool word, you gotta admit that

File: components/sup/src/manager/user_config_watcher.rs
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Why eventually? Do you mean that the operation is asynchronous?

If so, I think I can explain it more explicitly.

File: components/sup/src/manager/service/mod.rs
Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: Not sure what this "what?" is about, but to be clear - I don't actually want to have variables like `cufr` or `sucu`. I wrote it like that to avoid typing in the comment.

Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: I think that even with more descriptive names, it's hard to parse logically. There's just too much happening on one line: returning a tuple, ORing two values, calling functions..

Not sure it's an improvement.

Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: You don't need to put everything in one line. You could very well have:
```
let config_changed = config_updated_from_rumors || self.user_config_updated;
let needs_reload = self.compile_hooks(&ctx) || self.user_config_updated;
let needs_reconfigure = self.compile_configuration(&ctx);
(config_changed, needs_reload, needs_reconfigure)
```

In general, what I wanted to propose is to simplify the flow of the code, instead of using nested `if`s and `match`es that are sometimes executed and sometimes not.

Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: In the above implementation though, we're calling `self.compile_hooks` even if we don't need to, which will lead to "Hooks recompiled" being printed over and over (this function gets executed [on every tick](https://github.com/kinvolk/habitat/blob/0a1d210328174ae822c13213e2462f6f37f1e94c/components/sup/src/manager/service/mod.rs#L294)).

I'm definitely open to simplifying the code but obviously it has to have the same behaviour.

The problem is that some of these functions have side effects so we can't just call them willy-nilly.

Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: The `self.compile_hooks(&ctx)` should only be executed when either `cfg_updated_from_rumors` is true or if `census_ring.changed()` returns true. So if this happens everytime, then something is wrong.

**Comment by asymmetric** [comment]
Closing in favor of https://github.com/habitat-sh/habitat/pull/3898.

**Pull request closed without merging**
